File: csharp/Auction-TestDesign-Kata/Auction/AuctionMessageTranslator.cs
namespace Auction;

public class AuctionMessageTranslator(IAuctionEventListener listener)
{
    public void ProcessMessage(string message)
    {
        if (message.Contains("CLOSE"))
        {
            // bug: should notify listener
        }
        else if (message.Contains("PRICE"))
        {
            var data = ParseMessageData(message);
            
            var currentPrice = int.Parse(data["CurrentPrice"]);
            var increment = int.Parse(data["Increment"]);
            var bidder = data["Bidder"];

            // bug: should notify listener
        }
        else
        {
            // bug: should notify listener
        }
    }

    private static Dictionary<string, string> ParseMessageData(string message)
    {
        var data = new Dictionary<string, string>();
        foreach (var element in message.Split(";", StringSplitOptions.RemoveEmptyEntries))
        {
            var pair = element.Split(":");
            data[pair[0].Trim()] = pair[1].Trim();
        }

        return data;
    }
}

File: typescript/src/AuctionMessageTranslator.ts
import { AuctionEventListener } from './AuctionEventListener';

export class AuctionMessageTranslator {
    private readonly listener: AuctionEventListener;

    constructor(listener: AuctionEventListener) {
        this.listener = listener;
    }

    public processMessage(message: string) {
        if (message.includes('CLOSE')) {
            // bug: should notify listener
        } else if (message.includes('PRICE')) {
            const data: { [key: string]: string } = {};
            message
                .split(';')
                .filter((element) => element.includes(':'))
                .forEach((element) => {
                    const pair = element.split(':');
                    data[pair[0].trim()] = pair[1].trim();
                });

            const currentPrice = parseInt(data['CurrentPrice']);
            const increment = parseInt(data['Increment']);
            const bidder = data['Bidder'];

            // bug: should notify listener
        } else {
            // bug: should notify listener
        }
    }
}
